hudi-client/hudi-client-common/src/main/java/org/apache/hudi/async/HoodieAsyncTableService.java

```diff
@@ -47,4 +60,11 @@ public void start(Function<Boolean, Boolean> onShutdownCallback) {
     }
     super.start(onShutdownCallback);
   }
+
+  public void updateWriteConfig(HoodieWriteConfig writeConfig) {
+    synchronized (writeConfigUpdateLock) {
+      this.writeConfig = EmbeddedTimelineServerHelper.updateWriteConfigWithTimelineServer(embeddedTimelineService.get(), writeConfig);
+      isWriteConfigUpdated.set(true);
```

Comment:
Flag setup is quite brittle: it makes resetting a 2-step process (first update the flag, then update needs to be handled asynchronously), which is easy to miss during impl.

Instead i'd suggest we 
 - Create abstract hook `handleWriteConfigUpdate` (hence make every inheritor implement it)
 - Invoke it here under lock (that way we keep the lock private and avoid this logic spilling into inheritors)

Reply:
let's sync up. I am not sure if I understood your comment fully.
we can't update the writeConfig synchronously. The crux here is that, we need to re-instantiate a new writeClient if writeconfig is updated. but, we can give up on entire locking since all we care is, next time when compaction runs it should re-instate itself w/ updated write config.

Reply:
Sync'd up offline: we can reset the write-client (it won't affect already running compaction/clustering) but we won't be able to close existing client synchronously, and will have to do an async follow-up anyway

Reply:
@nsivabalan let's do the following instead:

 - Let's keep the flag to trigger the refresh
 - Let's make the flag more generic (like `shouldResetWriteClient`) so that it's not coupled to writeConfig (there's no need for it to be)
 - Let's reuse this logic b/w Clustering/Compaction services (see my comment above)